pkg/services/authz/rbac/resolver.go

```diff
@@ -0,0 +1,141 @@
+	return teamIDs, nil
+}
+
+// Should return an error if we fail to build the resolver.
+func (s *Service) newTeamNameResolver(ctx context.Context, ns types.NamespaceInfo) (ScopeResolverFunc, error) {
+	teamIDs, cacheHit := s.teamIDCache.Get(ctx, teamIDsCacheKey(ns.Value))
+	if !cacheHit {
+		var err error
+		teamIDs, err = s.fetchTeams(ctx, ns)
+		if err != nil {
+			return nil, fmt.Errorf("could not build resolver: %w", err)
+		}
+	}
+
+	return func(scope string) (string, error) {
+		teamIDStr := strings.TrimPrefix(scope, "teams:id:")
+		if teamIDStr == "" {
+			return "", fmt.Errorf("team ID is empty")
+		}
+		if teamIDStr == "*" {
+			return "teams:uid:*", nil
+		}
+		teamID, err := strconv.ParseInt(teamIDStr, 10, 64)
+		if err != nil {
+			return "", fmt.Errorf("invalid team ID %s: %w", teamIDStr, err)
+		}
+		if teamName, ok := teamIDs[teamID]; ok {
+			return "teams:uid:" + teamName, nil
+		}
+
+		// Stale cache recovery: Try to fetch the teams again.
+		if cacheHit {
+			// Potential future improvement: if multiple threads have the same stale cache,
+			// they might refetch teams separately and asynchronously. We could use a more sophisticated
+			// approach to avoid this. Like checking if the cache has been updated meanwhile.
+			cacheHit = false
+			teamIDs, err = s.fetchTeams(ctx, ns)
+			if err != nil {
+				// Other improvement: Stop the calling loop if we fail to fetch teams.
+				return "", err
+			}
+			if teamName, ok := teamIDs[teamID]; ok {
+				return "teams:uid:" + teamName, nil
+			}
+		}
+
+		return "", fmt.Errorf("team ID %s not found", teamIDStr)
+	}, nil
+}
+
+func (s *Service) nameResolver(ctx context.Context, ns types.NamespaceInfo, scopePrefix string) (ScopeResolverFunc, error) {
+	if strings.HasPrefix(scopePrefix, "teams:id:") {
+		return s.newTeamNameResolver(ctx, ns)
+	}
+	// No resolver found for the given scope prefix.
+	return nil, nil
+}
+
+// resolveScopeMap translates scopes like "teams:id:1" to "teams:uid:t1".
+// It assumes only one scope resolver is needed for a given scope map, based on the first valid scope encountered.
+func (s *Service) resolveScopeMap(ctx context.Context, ns types.NamespaceInfo, scopeMap map[string]bool) (map[string]bool, error) {
+	var (
+		prefix        string
+		scopeResolver ScopeResolverFunc
+		err           error
+	)
+	for scope := range scopeMap {
+		// Find the resolver based on the first scope with a valid prefix
+		if prefix == "" {
+			if len(strings.Split(scope, ":")) < 3 {
+				// Skip scopes that don't have at least 3 parts (e.g., "*", "teams:*")
+				// This is because we expect scopes to be in the format "resource:attribute:value".
+				continue
+			}
+			if strings.HasPrefix(scope, "folders:uid:") {
+				// Resources can be stored in folders
+				// but we don't want to resolve folder scopes
+				continue
+			}
+
+			// Initialize the scope resolver only once
+			prefix = accesscontrol.ScopePrefix(scope)
+			scopeResolver, err = s.nameResolver(ctx, ns, prefix)
```

Comment:
nit: why do we need to pass in the prefix and not the whole scope, if we then check `strings.HasPrefix(scopePrefix, "teams:id:")` anyway?